types/part_set.go

```diff
@@ -274,3 +279,74 @@ func (ps *PartSet) StringShort() string {
 		return fmt.Sprintf("(%v of %v)", ps.Count(), ps.Total())
 	}
 }
+
+//--------------------------------------
+// for gossiping parts rarest first
+
+// TODO: access individual counts atomically instead
+type PartCounts struct {
+	mtx    sync.Mutex
+	counts []uint32
+
+	lowestIndices []int // allocated once, used to randomly choose one of the lowest indices each time
+}
+
+func NewPartCounts(size int) *PartCounts {
+	return &PartCounts{
+		counts:        make([]uint32, size),
+		lowestIndices: make([]int, size),
+	}
+}
+
+// check() checks that the peer is hearing about the part for the first time.
+// the check/update must happen atomically on PartCounts since peer parts
+// are updated concurrently
+func (pc *PartCounts) CheckAndIncrementIndex(index int, check func() bool) {
+	pc.mtx.Lock()
+	defer pc.mtx.Unlock()
+	if check() {
+		pc.counts[index] += 1
+	}
+}
+
+// panics if index >= len(pc.counts)
+func (pc *PartCounts) IncrementIndex(index int) {
+	pc.mtx.Lock()
+	defer pc.mtx.Unlock()
+	pc.counts[index] += 1
+}
+
+// Pick the first index from the bit array with the lowest count, increment the index,
+// and pass it to the setFunc.
+// Index must be incremented so that multiple concurrent calls don't pick the same value,
+// and passed to setFunc so the abundance doesnt get accidentally incremented again
+// Assumes possibleParts.Size() == len(pc.counts)
+func (pc *PartCounts) PickRarest(possibleParts *cmn.BitArray, setFunc func(int)) int {
```

Comment:
can't we use here some tree (like binary heap) where the rarest element would always be on top?

Reply:
yep believe so!